node-api: deprecate napi_module_register based registration

vmoroz opened the pull request:
### Summary

Change `NAPI_MODULE` and `NAPI_MODULE_INIT` macros to use symbol-based module registration.
Symbol-based registration was added in PR #20161 and used by Node-API modules written in non-C++ languages.
It is much simpler and requires no work on module library load.

### Details

Currently ABI-safe Node-API modules have two ways to register modules:
- Create `napi_module` struct and call `napi_module_register` on module library (.so or .dll) load.
- Define `napi_register_module_v1` function which is called on-demand to initialize module exports.

Node.JS module binding loads module library and then
- if it finds registered `napi_module` struct, then it uses its `nm_register_func` to initialize module exports,
- otherwise, it tries to find the `napi_register_module_v1` and calls it to initialize module exports.

The first registration method is the original Node-API (NAPI) module registration. The second method (aka symbol-based registration) was added in PR #20161 on Apr 23, 2018, and available in Node.JS from version 10.

It is interesting to see that all [Node-API bindings](https://github.com/nodejs/abi-stable-node/blob/doc/node-api-engine-bindings.md) written in other languages (Rust, Go, etc.) they all use the symbol-based registration. Even the C/C++ WASM bindings use the symbol-based registration. It is easier to implement, and it does not need calling code on module library load.

In this PR we align C/C++ Node-API registration with registration in other languages to use the symbol-based registration. The `NAPI_MODULE` and `NAPI_MODULE_INIT` macros are changed to define the `napi_register_module_v1` function instead of calling `napi_module_register`. It effectively deprecates the `napi_module_register` function, but since no changes are made to the Node.JS binding mechanism, all previously compiled modules must continue to work.

The `test_null_init` is removed because new macros fail to compile if the `NULL` parameter is given to the `NAPI_MODULE` macro.

This PR is a "spin off" from the PR #45715 to keep it focused on the new versioning approach.

mhdawson commented:
@vmoroz I had thought when we discussed this that we'd not remove anything that could break existing code. This seems to remove  NAPI_MODULE_X which might do that?

vmoroz commented:
> @vmoroz I had thought when we discussed this that we'd not remove anything that could break existing code. This seems to remove NAPI_MODULE_X which might do that?

@mhdawson, you are right, I initially wanted to have just a small incremental change, but then went a bit overboard with deleting the old code. I could not find any code that currently uses `NAPI_MODULE_X` and wanted to hear feedback from you and other developers. Let me change it back to a less aggressive change that keeps all intermediate macros unchanged.

mhdawson commented:
@vmoroz I think we could do this, but I was thinking we might update  NAPI_MODULE_X to use the symbol as well. That would avoid breaking existing code although it would remove some information that can be passed in,however, I think we believed that info was not used/avaiable anywhere?

vmoroz commented:
> @vmoroz I think we could do this (ie just update NAPI_MODULE and leave NAPI_MODULE_X alone), but I was thinking we might update NAPI_MODULE_X to use the symbol as well versus removing it as was in an earlier verion. That would avoid breaking existing code although it would remove some information that can be passed in. However, I think we believed that info was not used/avaiable anywhere right?

@mhdawson , we can do it too. I see that `NAPI_MODULE_X` was added in the very first [Node-API PR](https://github.com/nodejs/node/pull/11975), but it is not clear if it is used. Do we have contacts from Electron team who can say if this macro is important for their scenarios?

vmoroz commented:
@mhdawson I have restored the original change in https://github.com/nodejs/node/pull/46319/commits/df4860982f04a2d4f6af86377871a963392d0784 and added the `NAPI_MODULE_X` macro.

Looking through the code https://github.com/nodejs/node/blob/8ba54e50496a6a5c21d93133df60a9f7cb6c46ce/src/node_api.cc#L622-L631 I can confirm that we never used the `NAPI_MODULE_X`'s `priv` parameter. This function is the registration callback used by the `napi_module_register`. We see that it calls the `napi_module_register_by_symbol` method used by the symbol-based registration.
As for the `flags` parameter, we only have four flags: https://github.com/nodejs/node/blob/8ba54e50496a6a5c21d93133df60a9f7cb6c46ce/src/node_binding.h#L15-L20 and the flag `NM_F_DELETEME` is always forced in https://github.com/nodejs/node/blob/8ba54e50496a6a5c21d93133df60a9f7cb6c46ce/src/node_api.cc#L682-L694 It means that the `NAPI_MODULE_X`'s `flags` parameter was never used too.

I added a comment for the `NAPI_MODULE_X` macro that it is deprecated and `NAPI_MODULE` must be used instead.

mhdawson reviewed (approved):
LGTM

mhdawson commented:
@vmoroz the changes to the main file look ok, I'm wondering why a test needs to be removed if we are not breaking anything?

vmoroz commented:
> @vmoroz the changes to the main file look ok, I'm wondering why a test needs to be removed if we are not breaking anything?

@mhdawson , the test was verifying behavior for what happens when we pass `NULL` function pointer to the `NAPI_MODULE` macro. The test was expecting to have a runtime error in such case.

The new `NAPI_MODULE` macro uses provided function to implement the `napi_register_module_v1` function. In case the provided function is `NULL` or any other invalid data, we get a compilation error. Thus, the test is not needed anymore and it cannot be compiled with the new macro.

legendecas reviewed (commented):
I think the test_null_init is still applicable as we didn't remove the method to register with `napi_module_register`, which can still be called with a null initializer, like when addons are built with old `node_api.h` header files.

vmoroz commented:
> I think the test_null_init is still applicable as we didn't remove the method to register with `napi_module_register`, which can still be called with a null initializer, like when addons are built with old `node_api.h` header files.

I have restored the test. It is important to note that since the test was relying on the old `NAPI_MODULE` macro I had to copy most of the old registration code into this test.

mhdawson commented:
@addaleax with the more recent changes I don't think this is SemVer major any more. Is it ok with you if I remove the label?

addaleax reviewed (commented):
@mhdawson If you’re prefer to land this as semver-minor, yeah, I think so :+1:

mhdawson commented on src/node_api.h:
```diff
@@ -116,11 +62,19 @@ typedef struct napi_module {
 EXTERN_C_START
 
+// Deprecated. Replaced by symbol-based registration defined by NAPI_MODULE
+// and NAPI_MODULE_INIT macros.
```
@vmoroz have you looked at this suggestion?

vmoroz replied:
@mhdawson , yes, it is part of the new code. Please see the lines 78-80 in the node_api.h below.
So far I had addressed all the previous PR feedback. I can rebase the code to be on top of the latest code.

mhdawson reviewed (approved):
LGTM

legendecas commented on src/node_api.h:
```diff
@@ -31,6 +31,7 @@ struct uv_loop_s;  // Forward declaration.
 typedef napi_value(NAPI_CDECL* napi_addon_register_func)(napi_env env,
                                                          napi_value exports);
 
+// Used by deprecated registration method napi_module_register.
```
Yeah, `AddLinkedBinding()` should be listed here too.

vmoroz replied:
@legendecas , the `AddLinkedBinding()` is not a Node-API API. Do we really want to list here all possible users of that struct even if they are not part of the Node-API? There quite a few internal functions that do it. My preference would be not to mention anything which is not part of the Node-API in this header.

legendecas reviewed (approved):
LGTM % one nit.

mhdawson commented:
Landed in 5e4f5e122c06
